src/query/query_hash_scan.c

```diff
@@ -506,12 +666,2483 @@ int
+		      er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_EH_CORRUPTED, 0);
+		      return false;
+		    }
+
+		  bucket_record_p = (char *) recdes.data;
+		  bucket_record_p += sizeof (OID);
+
+		  if (fhs_compare_key (thread_p, bucket_record_p, key_type, key_p, recdes.type, &compare_result) !=
+		      NO_ERROR)
+		    {
+		      return false;
+		    }
+		}
+	      *out_position_p = middle + 1;
+	    }
+	  else
+	    {
+	      /* go to last slot using flag */
+	      if (spage_get_record (thread_p, bucket_page_p, middle, &recdes, PEEK) != S_SUCCESS)
+		{
+		  er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_EH_CORRUPTED, 0);
+		  return false;
+		}
+	      fhs_read_flag_from_record ((char *) recdes.data, &flag);
+	      if (flag == FHS_FLAG_INDIRECT)
+		{
+		  *out_position_p = middle;
+		}
+	      else if (flag >= 1)
+		{
+		  *out_position_p = middle + flag - 1;
+		}
+	      else
+		{
+		  er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_EH_CORRUPTED, 0);
+		  return false;
+		}
+	    }
+	  return true;
+	}
+
+      if (compare_result < 0)
+	{
+	  high = middle - 1;
+	}
+      else
+	{
+	  low = middle + 1;
+	}
+    }
+  while (high >= low);
+
+  if (high < middle)
+    {
+      *out_position_p = middle;
+    }
+  else
+    {
+      /* Key is NOT in bucket; it should be located at the right of middle key. */
+      *out_position_p = middle + 1;
+    }
+
+  return false;
+}
+
+/*
+ * fhs_insert () - Insert (key, assoc_value) pair to ext. hashing
+ *   return: void * (NULL is returned in case of error)
+ *   fhsid(in): identifier of the extendible hashing structure
+ *   key(in): key value to insert
+ *   value_ptr(in): pointer to the associated value to insert
+ *
+ * Note: Inserts the given (key & assoc_value) pair into the specified
+ * extendible hashing structure. If the key already exists then
+ * the previous associated value is replaced with the new one.
+ * Otherwise, a new record is inserted to the correct bucket.
+ *      To perform the insertion operation safely in the concurrent
+ * environment, an auxiliary (and private) function
+ */
+void *
+fhs_insert (THREAD_ENTRY * thread_p, FHSID * fhsid_p, void *key_p, OID * value_p)
+{
+  VPID bucket_vpid;
+
+  FHS_HASH_KEY hash_key;
+  int location;
+  FHS_RESULT result;
+
+  if (fhs_find_bucket_vpid_with_hash (thread_p, fhsid_p, key_p,
+				      PGBUF_LATCH_WRITE, PGBUF_LATCH_READ, &bucket_vpid, &hash_key,
+				      &location) != NO_ERROR)
+    {
+      return NULL;
+    }
+
+  if (VPID_ISNULL (&bucket_vpid))
+    {
+      if (fhs_insert_to_bucket_after_create (thread_p, fhsid_p, &bucket_vpid, location, hash_key, key_p, value_p) !=
+	  NO_ERROR)
+	{
+	  return NULL;
+	}
+    }
+  else
+    {
+      result = fhs_insert_bucket_after_extend_if_need (thread_p, fhsid_p, &bucket_vpid, key_p, hash_key, value_p);
+      if (result == FHS_ERROR_OCCURRED)
+	{
+	  return NULL;
+	}
+      else if (result == FHS_BUCKET_FULL)
+	{
+	  return fhs_insert (thread_p, fhsid_p, key_p, value_p);
```

Comment:
If bucket is full, bucket is extended in `fhs_insert_bucket_after_extend_if_need()`. Can I know why fhs_insert should be called recursively in this case?

Reply:
This is to split the bucket again.
This is occurred when data is inserted into only one of the split buckets.